File: app_tests/elixir/vulnerable_app.ex
defmodule VulnerableApp do
  @moduledoc """
  A vulnerable Elixir application demonstrating common security issues
  """

  # Hardcoded secrets
  @secret_key "super_secret_elixir_key_2023"
  @api_token "elx_1234567890abcdef"
  @database_password "admin123!@#"

  def main do
    IO.puts("=== Vulnerable Elixir Application ===")
    IO.puts("Secret Key: #{@secret_key}")
    
    IO.puts("\n1. Testing atom exhaustion vulnerability...")
    atom_exhaustion_demo()
    
    IO.puts("\n2. Testing code injection...")
    code_injection_demo()
    
    IO.puts("\n3. Testing unsafe deserialization...")
    unsafe_deserialization_demo()
    
    IO.puts("\n4. Testing path traversal...")
    path_traversal_demo()
    
    IO.puts("\n5. Testing weak randomness...")
    weak_randomness_demo()
    
    IO.puts("\n6. Testing information disclosure...")
    information_disclosure_demo()
  end

  def atom_exhaustion_demo do
    IO.write("Enter dynamic atom name: ")
    user_input = IO.read(:stdio, :line) |> String.trim()
    
    # Atom exhaustion vulnerability - converts user input directly to atom
    # This can exhaust the atom table and crash the VM
    dangerous_atom = String.to_atom(user_input)
    IO.puts("Created atom: #{dangerous_atom}")
  rescue
    e -> IO.puts("Atom creation error: #{inspect(e)}")
  end

  def code_injection_demo do
    IO.write("Enter Elixir code to evaluate: ")
    user_code = IO.read(:stdio, :line) |> String.trim()
    
    # Code injection vulnerability - evaluates user input as code
    try do
      {result, _} = Code.eval_string(user_code)
      IO.puts("Code result: #{inspect(result)}")
    rescue
      e -> IO.puts("Code evaluation error: #{inspect(e)}")
    end
  end

  def unsafe_deserialization_demo do
    IO.write("Enter Erlang term to deserialize: ")
    user_input = IO.read(:stdio, :line) |> String.trim()
    
    # Unsafe deserialization - directly deserializes user input
    # This can lead to code execution if malicious terms are provided
    try do
      term = :erlang.binary_to_term(user_input)
      IO.puts("Deserialized term: #{inspect(term)}")
    rescue
      e -> IO.puts("Deserialization error: #{inspect(e)}")
    end
  end

  def path_traversal_demo do
    IO.write("Enter file path to read: ")
    file_path = IO.read(:stdio, :line) |> String.trim()
    
    # Path traversal vulnerability - no validation of file path
    case File.read(file_path) do
      {:ok, content} ->
        content_preview = String.slice(content, 0, 200)
        IO.puts("File content: #{content_preview}...")
      {:error, reason} ->
        IO.puts("File read error: #{reason}")
    end
  end

  def weak_randomness_demo do
    # Weak randomness - predictable random numbers
    :rand.seed(:exrop, {1, 2, 3})  # Fixed seed
    weak_token = :rand.uniform(1000000)
    IO.puts("Predictable session token: #{weak_token}")
    
    # Using current time as seed (predictable)
    time_seed = System.system_time(:nanosecond)
    :rand.seed(:exrop, {time_seed, 0, 0})
    time_based_token = :rand.uniform(1000000)
    IO.puts("Time-based token: #{time_based_token}")
  end

  def information_disclosure_demo do
    try do
      # Simulating a database connection error
      raise "Database connection failed: host=localhost, user=admin, password=#{@database_password}"
    rescue
      e ->
        # Information disclosure through error messages
        IO.puts("Full error with sensitive info: #{Exception.message(e)}")
        IO.puts("Stack trace: #{Exception.format_stacktrace(__STACKTRACE__)}")
        IO.puts("API Token for debugging: #{@api_token}")
    end
  end

  def sql_injection_simulation do
    IO.write("Enter user ID: ")
    user_id = IO.read(:stdio, :line) |> String.trim()
    
    # SQL injection vulnerability (simulated)
    # In a real app, this would be passed to a database query
    query = "SELECT * FROM users WHERE id = #{user_id}"
    IO.puts("Executing vulnerable query: #{query}")
    
    # Command injection through system calls
    IO.write("Enter filename to process: ")
    filename = IO.read(:stdio, :line) |> String.trim()
    
    # Command injection vulnerability
    command = "ls -la #{filename}"
    IO.puts("Executing command: #{command}")
    
    try do
      {output, exit_code} = System.cmd("sh", ["-c", command])
      IO.puts("Command output: #{output}")
      IO.puts("Exit code: #{exit_code}")
    rescue
      e -> IO.puts("Command execution error: #{inspect(e)}")
    end
  end

  def unsafe_process_operations do
    IO.write("Enter process name to start: ")
    process_name = IO.read(:stdio, :line) |> String.trim()
    
    # Unsafe process spawning with user input
    pid = spawn(fn ->
      # This could be dangerous if process_name contains malicious code
      apply(String.to_atom(process_name), :start, [])
    end)
    
    IO.puts("Started process: #{inspect(pid)}")
  rescue
    e -> IO.puts("Process spawn error: #{inspect(e)}")
  end

  def regex_dos_demo do
    IO.write("Enter text to match against complex regex: ")
    user_text = IO.read(:stdio, :line) |> String.trim()
    
    # ReDoS vulnerability - catastrophic backtracking
    vulnerable_regex = ~r/^(a+)+b$/
    
    start_time = System.monotonic_time(:millisecond)
    result = Regex.match?(vulnerable_regex, user_text)
    end_time = System.monotonic_time(:millisecond)
    
    IO.puts("Regex match result: #{result}")
    IO.puts("Time taken: #{end_time - start_time} ms")
  end

  # GenServer with state injection vulnerability
  defmodule VulnerableServer do
    use GenServer

    def start_link(_) do
      GenServer.start_link(__MODULE__, %{}, name: __MODULE__)
    end

    def init(state) do
      {:ok, state}
    end

    # Vulnerable: allows arbitrary state injection
    def handle_call({:update_state, new_state}, _from, _state) do
      {:reply, :ok, new_state}
    end

    def handle_call(:get_state, _from, state) do
      {:reply, state, state}
    end
  end
end

# Start the application
VulnerableApp.main()

File: app_tests/erlang/vulnerable_app.erl
-module(vulnerable_app).
-export([main/0, start/0]).

% Hardcoded secrets
-define(SECRET_KEY, "erlang_super_secret_2023").
-define(API_TOKEN, "erl_1234567890abcdef").
-define(DB_PASSWORD, "admin123!@#").

main() ->
    io:format("=== Vulnerable Erlang Application ===~n"),
    io:format("Secret Key: ~s~n", [?SECRET_KEY]),
    
    io:format("~n1. Testing atom exhaustion vulnerability...~n"),
    atom_exhaustion_demo(),
    
    io:format("~n2. Testing code injection...~n"),
    code_injection_demo(),
    
    io:format("~n3. Testing unsafe deserialization...~n"),
    unsafe_deserialization_demo(),
    
    io:format("~n4. Testing command injection...~n"),
    command_injection_demo(),
    
    io:format("~n5. Testing weak randomness...~n"),
    weak_randomness_demo(),
    
    io:format("~n6. Testing information disclosure...~n"),
    information_disclosure_demo(),
    
    ok.

start() ->
    main().

atom_exhaustion_demo() ->
    io:format("Enter dynamic atom name: "),
    {ok, [UserInput]} = io:fread("", "~s"),
    
    try
        % Atom exhaustion vulnerability - converts user input directly to atom
        % This can exhaust the atom table and crash the VM
        DangerousAtom = list_to_atom(UserInput),
        io:format("Created atom: ~w~n", [DangerousAtom])
    catch
        Error:Reason ->
            io:format("Atom creation error: ~w:~w~n", [Error, Reason])
    end.

code_injection_demo() ->
    io:format("Enter Erlang expression to evaluate: "),
    {ok, [UserCode]} = io:fread("", "~s"),
    
    try
        % Code injection vulnerability - evaluates user input as Erlang code
        {ok, Tokens, _} = erl_scan:string(UserCode ++ "."),
        {ok, ParsedExpr} = erl_parse:parse_exprs(Tokens),
        {value, Result, _} = erl_eval:exprs(ParsedExpr, []),
        io:format("Code result: ~w~n", [Result])
    catch
        Error:Reason ->
            io:format("Code evaluation error: ~w:~w~n", [Error, Reason])
    end.

unsafe_deserialization_demo() ->
    io:format("Enter base64 encoded term to deserialize: "),
    {ok, [UserInput]} = io:fread("", "~s"),
    
    try
        % Unsafe deserialization - directly deserializes user input
        % This can lead to code execution if malicious terms are provided
        DecodedBinary = base64:decode(UserInput),
        Term = binary_to_term(DecodedBinary),
        io:format("Deserialized term: ~w~n", [Term])
    catch
        Error:Reason ->
            io:format("Deserialization error: ~w:~w~n", [Error, Reason])
    end.

command_injection_demo() ->
    io:format("Enter filename to list: "),
    {ok, [FileName]} = io:fread("", "~s"),
    
    % Command injection vulnerability
    Command = "ls -la " ++ FileName,
    io:format("Executing command: ~s~n", [Command]),
    
    try
        Result = os:cmd(Command),
        io:format("Command output: ~s~n", [Result])
    catch
        Error:Reason ->
            io:format("Command execution error: ~w:~w~n", [Error, Reason])
    end.

weak_randomness_demo() ->
    % Weak randomness - predictable random numbers
    random:seed(1, 2, 3),  % Fixed seed (deprecated but still vulnerable)
    WeakToken = random:uniform(1000000),
    io:format("Predictable session token: ~w~n", [WeakToken]),
    
    % Using current time as seed (predictable)
    {A, B, C} = now(),
    random:seed(A, B, C),
    TimeBasedToken = random:uniform(1000000),
    io:format("Time-based token: ~w~n", [TimeBasedToken]).

information_disclosure_demo() ->
    try
        % Simulating a database connection error
        error({database_connection_failed, 
               [{host, "localhost"}, 
                {user, "admin"}, 
                {password, ?DB_PASSWORD}]})
    catch
        Error:Reason ->
            % Information disclosure through error messages
            io:format("Full error with sensitive info: ~w:~w~n", [Error, Reason]),
            io:format("Stack trace: ~p~n", [erlang:get_stacktrace()]),
            io:format("API Token for debugging: ~s~n", [?API_TOKEN])
    end.

% Vulnerable process with state injection
vulnerable_server() ->
    receive
        {update_state, NewState} ->
            % Vulnerable: allows arbitrary state injection
            io:format("State updated to: ~w~n", [NewState]),
            vulnerable_server_loop(NewState);
        {get_state, From} ->
            From ! {state, undefined},
            vulnerable_server()
    end.

vulnerable_server_loop(State) ->
    receive
        {update_state, NewState} ->
            io:format("State updated from ~w to ~w~n", [State, NewState]),
            vulnerable_server_loop(NewState);
        {get_state, From} ->
            From ! {state, State},
            vulnerable_server_loop(State);
        {execute, Code} ->
            % Dangerous: executes arbitrary code
            try
                {ok, Tokens, _} = erl_scan:string(Code ++ "."),
                {ok, ParsedExpr} = erl_parse:parse_exprs(Tokens),
                {value, Result, _} = erl_eval:exprs(ParsedExpr, []),
                io:format("Executed code result: ~w~n", [Result])
            catch
                Error:Reason ->
                    io:format("Code execution error: ~w:~w~n", [Error, Reason])
            end,
            vulnerable_server_loop(State)
    end.

% SQL injection simulation (would be used with a database driver)
sql_injection_simulation() ->
    io:format("Enter user ID: "),
    {ok, [UserId]} = io:fread("", "~s"),
    
    % SQL injection vulnerability (simulated)
    Query = "SELECT * FROM users WHERE id = " ++ UserId,
    io:format("Executing vulnerable query: ~s~n", [Query]),
    
    % In a real application, this would be passed to a database
    ok.

% Buffer overflow simulation using binary operations
buffer_overflow_simulation() ->
    io:format("Enter data to process: "),
    {ok, [UserData]} = io:fread("", "~s"),
    
    % Simulated buffer overflow - processing user data without bounds checking
    SmallBuffer = binary:part(list_to_binary(UserData), 0, min(length(UserData), 64)),
    LargeData = list_to_binary(UserData),
    
    io:format("Small buffer size: ~w~n", [byte_size(SmallBuffer)]),
    io:format("Large data size: ~w~n", [byte_size(LargeData)]),
    
    % Potential issue: assuming buffer is always smaller
    if 
        byte_size(LargeData) > 64 ->
            io:format("WARNING: Data larger than buffer, potential overflow!~n");
        true ->
            io:format("Data fits in buffer~n")
    end.

% Race condition vulnerability
race_condition_demo() ->
    SharedState = spawn(fun() -> shared_state_process(0) end),
    
    % Spawn multiple processes that modify shared state without synchronization
    spawn(fun() -> 
        timer:sleep(100),
        SharedState ! {increment, self()},
        receive Msg -> io:format("Process 1 got: ~w~n", [Msg]) end
    end),
    
    spawn(fun() -> 
        timer:sleep(100),
        SharedState ! {increment, self()},
        receive Msg -> io:format("Process 2 got: ~w~n", [Msg]) end
    end),
    
    timer:sleep(500).

shared_state_process(State) ->
    receive
        {increment, From} ->
            % Race condition: read-modify-write without proper synchronization
            NewState = State + 1,
            timer:sleep(10),  % Simulate some processing time
            From ! {state, NewState},
            shared_state_process(NewState);
        {get_state, From} ->
            From ! {state, State},
            shared_state_process(State)
    end.
